table/tables/tables.go

```diff
@@ -399,15 +399,20 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx sessionctx.Context,
 			binlogNewRow = append(binlogNewRow, value)
 		}
 	}
+	sessVars := sctx.GetSessionVars()
 
 	// rebuild index
+	savePresumeKeyNotExist := sessVars.PresumeKeyNotExists
+	if !sessVars.ConstraintCheckInPlace {
+		sessVars.PresumeKeyNotExists = true
```

Comment:
I guess it should work under a pessimistic transaction.

https://github.com/pingcap/tidb/blob/master/session/txn.go#L288

I mean, if we presume a key does not exist, we will not write the index KV entry to TiKV immediately, but we should still **lock the KV entry to make pessimistic work as expected.**


So it's better for this PR to come with a test case:

s1: set @@tidb_txn_mode = 'pessimistic';
s1: set @@tidb_constraint_check_in_place = 0;
s2: set @@tidb_txn_mode = 'pessimistic';
s1: begin; update t set k = 3 where id = 2;
s2: insert into t values (2, 3, 'tidb');    // should block
s1: commit;

To check a block, this code snippet maybe help:

```
...
c := make(chan int, 2)
go func() {
    s2.exec("insert into t values (2, 3, 'tidb'))
    c <- 2
}
sleep(100ms)
c <- 1
s1.exec("commit")
assert(1 == <- c)
```

@xiaodong-ji

Reply:
done